### DISCUSSION TITLE
ZPT compatibility

### DISCUSSION THREAD

**Discussion by georgpfolz** [alpinejs/alpine#3874] [3. Feature ideas]
When using alpine.js with Zope Page Templates, all directives containing a colon provoke a compilation error `unknown namespace directive`.

For some directives, there is the possibility to just use the shorthand version as a workaround, like `x-on:click` → `@click`. Unfortunately there are directives whose shortcut also contains a colon, such as `:attributes`.

Fortunately there is a workaround in Zope, that's using `tal:attributes` to create the directive dynamically: `<button tal:attributes="x-on:click string:alert('')">alert</button>`. This works fine, but it makes the code somewhat less readable.

Maybe there could be an alternative to the colon that works with the alpine parser, some other character that's not used in any known programming context like `§` or `…` (U+2026), eg: `x-on§click=alert('')` or `x-on…click=alert('')`.

I also wouldn't mind having to specify the new separator explicitly, for example in `x-init`.

**Comment by ekwoka** [comment]
Currently it isn't supported or configurable.

So the only way would be to use x-bind to handle those that need this.

```html
:attributes="5"

// to

x-bind="{ attributes: 5 }"
```

Since it seems ZPT is a preprocessor or something, then implementing compatibility there is more sensible. Like providing it a list of directives that it doesn't handle itself. Since it could impact many different things, not just Alpine.

**Comment by georgpfolz** [comment]
I'm aware that it's not currently supported, I intended my post to be a feature request.

The list of directives ZPT can't handle is: all directives containing a colon. The part on the left of a colon is interpreted as namespace prefix in ZPT thus throwing an error (there is also an error if it's a shorthand starting with a colon, like `:attributes`).

Thank you for pointing out the `x-bind` possibility, that could be an alternative to using Zope's `tal:attributes` machinery.

I'm not really in an impasse here, since I can handle the problematic directives with `tal:attributes` just fine, but it surely would be more convenient to be able to use the directives in the html tags directly.

I know Zope is not so widespread anymore, but to my knowledge this kind of template syntax is used in other projects as well, such as [Chameleon](https://chameleon.readthedocs.io) or [PHPTAL](https://phptal.org), so it might be worthwile to have a solution to make alpine more compatible with it.

**Reply by ekwoka** [reply]
Seems like getting compatibility with things nobody uses that stopped being maintained 6+ years ago wouldn't be worth the effort (the others, I mean. I do see ZPT is at least maintained).

But I'm not the maintainer. You can open a PR though 👍 

Or maybe there is a way to add a namespace plugin to ZPT that just passes alpine stuff through.

**Reply by georgpfolz** [reply]
Zope ist quite active at the moment with the last update being from last week, and PHPTAL seems to have had its last update last month. 

Adding a namespace to Zope is unfortunately less trivial than I thought initially (it cannot be done just by configuring the site.zcml file of a Zope instance, as I hoped).

I will consider opening a PR! 🙂

**Reply by ekwoka** [reply]
You would think if it had an unknown namespace it would just let it pass through.
